### PR TITLE
Data Export

### PR DESCRIPTION
**Pull request by hector0513**
image

### CODE REVIEW

**Comment by hector0513** [comment]
working progress

File: src/components/Header/ListExport.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+function mapDispatchToProps(dispatch) {
+    return {
+
+    }
+  }
+function url(type,data){
+  let blob ={}
+  if(type==="json"){
+     blob = new Blob([JSON.stringify(data)],{type:""})
```
Comment (Connoropolous) [inline src/components/Header/ListExport.js]:
```suggestion
     blob = new Blob([JSON.stringify(data, null, 2)],{type:""})
```

by doing this, you get JSON with nice formatting :)

**Comment by Connoropolous** [comment]
@hector0513 we need to separate lists by type in the CSV format. 

agents
____

goals
____

etc. 


Screen Shot 2019-11-26 at 10 23 46 AM

**Comment by hector0513** [comment]
> @hector0513 we need to separate lists by type in the CSV format.
> 
> agents
> 
> goals
> 
> etc.
> 
> Screen Shot 2019-11-26 at 10 23 46 AM


image
i already fix it

**Comment by Connoropolous** [comment]
how do we tell the difference between table headers and values? 

I can tell the difference, but Sam, Pegah, or someone else likely couldn't. 

Can we add line breaks between tables?

**Comment by hector0513** [comment]
> ¿Cómo distinguimos la diferencia entre los encabezados de tabla y los valores?
> 
> Puedo notar la diferencia, pero Sam, Pegah u otra persona probablemente no podrían.
> 
> ¿Podemos agregar saltos de línea entre tablas?

Like this ?
image

**Comment by alex1812r** [comment]
image

File: src/components/ListExport/Popup/Popup.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import React from 'react'
+import './popup.css'
+
+export default function Popup({ active, handleToHide },props){
+  return(
+    <div className={'popup ' + (active ? 'active' : '')}>
+      <div className='popup-window'>
+      <button onClick={handleToHide} className='btn-close-modal'>
+        <img src="/img/x.svg"/>
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]: Please use the `<Icon />` component that already exists in the project

File: src/components/ListExport/Popup/Popup.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      </button>
+        <div className="popup-header">
+          <span className="popup-logo">
+            <img src="/img/export.svg" />
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]: And here

File: src/components/Header/Header.js [hunk 4]
```diff
@@ -41,25 +52,35 @@ class Header extends React.Component {
...
   clickAvatar(e) {
-    this.setState({ isProfileOpen: !this.state.isProfileOpen, isStatusOpen: false, isGuideOpen: false })
+    this.setState({ isProfileOpen: !this.state.isProfileOpen,isExportOpen:false, isStatusOpen: false, isGuideOpen: false })
   }
   hover(bool) {
     this.setState({ avatar: bool })
   }
+  hoverExport(bool) {
+    this.setState({ isHoverExport: bool });
+    console.log(bool)
```
Comment (Connoropolous) [inline src/components/Header/Header.js]:
```suggestion
```

File: src/components/Header/Header.js [hunk 5]
```diff
@@ -41,25 +52,35 @@ class Header extends React.Component {
 
       ],
       avatar: false,
-      listaProfile: [{ title: "Profile Settings", click: this.clickProfile }]
+      listaProfile: [{ title: "Profile Settings", click: this.clickProfile }],
+      listaExport: [{ title: "Export s JSON",type:"json", download:"table.json" },
```
Comment (Connoropolous) [inline src/components/Header/Header.js]:
```suggestion
      listaExport: [{ title: "Export as JSON",type:"json", download:"table.json" },
```

File: src/components/Header/Header.js [hunk 6]
```diff
@@ -17,21 +19,30 @@ class Header extends React.Component {
...
-      listaProfile: {}
+      isExportOpen:false,
+      listaProfile: {},
+      listaExport: {}
+
     }
   }
 
   componentDidMount() {
+    const data ="hola"
```
Comment (Connoropolous) [inline src/components/Header/Header.js]: ?

File: src/components/ListExport/ListExport.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    const loop = (heardes, data) => {
+      const csvRows = [];
+
+      csvRows.push(Object.keys(data[heardes[0]]).join(","));
+      for (const index in heardes) {
+        csvRows.push(Object.values(data[heardes[index]]));
+      }
+      return csvRows.join("\n");
+    };
+    if (agents.length > 0) csvRows.push(loop(agents, data.agents));
```
Comment (Connoropolous) [inline src/components/ListExport/ListExport.js]:
```suggestion
    if (agents.length > 0) csvRows.push(loop("agents", agents, data.agents));
```

File: src/components/ListExport/ListExport.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+}
+function url(type, data) {
+  let blob = {};
+  if (type === "csv") {
+    const csvRows = [];
+    const agents = Object.keys(data.agents);
+    const goals = Object.keys(data.goals);
+    const edges = Object.keys(data.edges);
+    const goalMembers = Object.keys(data.goalMembers);
+    const loop = (heardes, data) => {
```
Comment (Connoropolous) [inline src/components/ListExport/ListExport.js]:
```suggestion
    const loop = (dataset, heardes, data) => {
```

File: src/components/ListExport/ListExport.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      const csvRows = [];
+
+      csvRows.push(Object.keys(data[heardes[0]]).join(","));
```
Comment (Connoropolous) [inline src/components/ListExport/ListExport.js:32]:
```suggestion
      csvRows.push(dataset)
      csvRows.push(Object.keys(data[heardes[0]]).join(","));
```

File: src/components/ListExport/ListExport.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      for (const index in heardes) {
+        csvRows.push(Object.values(data[heardes[index]]));
+      }
+      return csvRows.join("\n");
+    };
+    if (agents.length > 0) csvRows.push(loop(agents, data.agents));
+    if (goals.length > 0) csvRows.push("\n" + loop(goals, data.goals));
```
Comment (Connoropolous) [inline src/components/ListExport/ListExport.js]:
```suggestion
    if (goals.length > 0) csvRows.push("\n" + loop("goals", goals, data.goals));
```

File: src/components/ListExport/ListExport.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    if (edges.length > 0) csvRows.push("\n" + loop(edges, data.edges));
```
Comment (Connoropolous) [inline src/components/ListExport/ListExport.js]:
```suggestion
    if (edges.length > 0) csvRows.push("\n" + loop("edges", edges, data.edges));
```

File: src/components/ListExport/ListExport.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    if (goalMembers.length > 0)
+      csvRows.push("\n" + loop(goalMembers, data.goalMembers));
```
Comment (Connoropolous) [inline src/components/ListExport/ListExport.js]:
```suggestion
      csvRows.push("\n" + loop("goalMembers", goalMembers, data.goalMembers));
```

**Review by Connoropolous — CHANGES_REQUESTED** [review]
So close to complete. Nice work

**Comment by Connoropolous** [comment]
This button in the popup doesn't work at the moment. 

Screen Shot 2019-11-28 at 10 57 04 AM

File: src/components/ListExport/Popup/Popup.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        </div>
+        <div className="popup-content">
+          <p>
+            You just exported the <b>H-BE SoA</b> canvas a JSON file. 
+            You will be able to find it in your Downloads folder!
+          </p>
+        </div>
+        <div className="popup-footer">
+          <div className="check-no-show-again">
+            <label htmlFor="">Dont't show me again</label>
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]:
```suggestion
            <label htmlFor="">Don't show me again</label>
```

File: src/components/ListExport/Popup/Popup.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+            <input type="checkbox"/>
+          </div>
+          <button className="btn-accept">
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]: Use our custom `<Button />` component here instead, and attach an `onClick` listener to it

**Review by Connoropolous — CHANGES_REQUESTED** [review]
only very small changes still left here now

File: src/components/ListExport/Popup/Popup.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import React from 'react'
+import './popup.css'
+
+export default function Popup({ active, handleToHide },props){
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]:
```suggestion
export default function Popup({ active, handleToHide }){
```

File: src/components/ListExport/Popup/Popup.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  return(
+    <div className={'popup ' + (active ? 'active' : '')}>
+      <div className='popup-window'>
+      <button onClick={handleToHide} className='btn-close-modal'>
+         <Icon name="/img/x.svg" />
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]: Icon needs to be imported. and remove the `button` wrapper and just attach `onClick` to this `Icon` element

Comment (Connoropolous) [reply src/components/ListExport/Popup/Popup.js]: `<Icon name="x.svg" onClick={handleToHide} className="btn-close-modal" />`

File: src/components/ListExport/Popup/Popup.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+          
+      </button>
+        <div className="popup-header">
+          <span className="popup-logo">
+              <Icon name="/img/export.svg"/>
```
Comment (Connoropolous) [inline src/components/ListExport/Popup/Popup.js]: `<Icon name="export.svg" />`

**Review by Connoropolous — APPROVED** [review]
Good job guys!

🎉 🎉 🎉

**Pull request merged**